[dnfdaemon] Repolist and repoinfo

m-blaha opened the pull request.

m-blaha committed:
[dnfdaemon-client] Basic infrastructure + repolist

This is basically copy of relevant parts of microdnf client code.

m-blaha committed:
[dnfdaemon] Repolist - server part

Dbus server support for listing the repositories.

m-blaha committed:
[dnfdaemon] Repoinfo - server part

m-blaha committed:
[dnfdaemon-client] Repoinfo command

m-blaha committed:
[dnfdaemon] Catch repository loading errors

m-blaha committed:
[dnfdaemon] Correctly join all created threads

m-blaha committed:
[dnfdaemon] Fix several deadlocks

m-blaha committed:
[dnfdaemon] Threads manager

New ThreadsManager class serves as a "container" for all background
tasks. Class provides method for registering new thread and eventually
joining all the threads.
All dbus handlers are run asynchronously.
All exceptions in worker threads are catchec.

m-blaha committed:
[dnfdaemon-client] Use ArgumentParser from libdnf::cli

m-blaha committed:
[dnfdaemon] Replace ifilter_reponame -> ifilter_repoid

m-blaha committed:
[dnfdaemon] Set correct required version of sdbus-cpp

jrohel commented on dnfdaemon-client/commands/repolist/repolist.cpp:
```diff
@@ -0,0 +1,189 @@
+class RepoDbus {
+public:
+    RepoDbus(dnfdaemon::KeyValueMap & rawdata) : rawdata(rawdata){};
+    std::string get_id() { return rawdata["id"]; }
```
The method will add item (with default value) to the map if it does not exist? Is it intended? What about to make the methods `const`?

Examples:
1. Will return default value if key does not exist (but value is not added to the map):
`std::string get_id() const { auto it = rawdata.find("name"); return it != rawdata.end() ? it->second : sdbus::Variant(); }`

2. Will generate exception if key does not exist:
`std::string get_id() const { return rawdata.at("id"); }`

jrohel commented on dnfdaemon-client/utils.cpp:
```diff
@@ -0,0 +1,196 @@
+/*
```
Who use functions from this file?

m-blaha replied:
Oh, thanks. It's a leftover of copying microdnf structure to dnfdaemon-client.

jrohel commented on dnfdaemon-client/utils.hpp:
```diff
@@ -0,0 +1,61 @@
+/*
```
Who use functions from this file?

jrohel commented on dnfdaemon-server/session.cpp:
```diff
@@ -48,29 +51,50 @@ ItemType Session::session_configuration_value(const std::string & key, const Ite
+    // set cachedir
+    auto system_cache_dir = base->get_config().system_cachedir().get_value();
+    base->get_config().cachedir().set(libdnf::Option::Priority::RUNTIME, system_cache_dir);
+    // set variables
```
Sync "set variables" with upstream code.
Maybe replace by:
```
    base->get_vars().load(
        base->get_config().installroot().get_value(),
        base->get_config().varsdir().get_value()
    );
```

jrohel commented on dnfdaemon-server/threads_manager.cpp:
```diff
@@ -0,0 +1,83 @@
+
+    for (auto thread = to_be_joined.begin(); thread < to_be_joined.end(); ++thread) {
+        // join the thread and remove it from registry
+        if (thread->joinable()) {
```
Why the test? There are only joinable threads in the vector (test on line 55). Or another thread can change the state in the meantime? But in this case there is still a race condition betwen the `joinable()` test and `join()`.

jrohel commented on dnfdaemon-client/commands/repolist/repolist.cpp:
```diff
@@ -0,0 +1,189 @@
+    dnfdaemon::KeyValueMap options = {};
+    options["enable_disable"] = enable_disable_option->get_value();
+    std::vector<std::string> patterns_to_show;
+    if (patterns_to_show_options->size() > 0) {
```
for better readability: `size > 0` means is not empty
`if (!patterns_to_show_options->empty()) {`

m-blaha committed:
[dnfdaemon] Minor fixes

- rebase to the latest upstream code and use vars loading code from
  there
- remove unused dnfdaemon-client/utils.[ch]pp

m-blaha commented:
@jrohel Thanks for the review. The code has been rebased to the latest master and all your comments should be addressed.
